File: lib/runtime/event-handler.ts
import type { ModelFamily } from "../prompts/codex.js";
import type { AccountStorageV3 } from "../storage.js";
import { handleAccountSelectEvent } from "./account-select-event.js";

type RuntimeSelectableAccount = AccountStorageV3["accounts"][number];
type RuntimeSelectionChangedPayload = {
	previousIndex: number;
	nextIndex: number;
	account: RuntimeSelectableAccount;
};

export function createRuntimeEventHandler<
	TLoadedStorage extends AccountStorageV3 | null,
	TSavedStorage extends AccountStorageV3,
	TModelFamily extends string,
>(deps: {
	handleAccountSelectEvent: (input: {
		event: { type: string; properties?: unknown };
		providerId: string;
		loadAccounts: () => Promise<TLoadedStorage>;
		saveAccounts: (storage: TSavedStorage) => Promise<void>;
		modelFamilies: readonly TModelFamily[];
		syncSelectedAccount?: (
			index: number,
			account: RuntimeSelectableAccount,
		) => Promise<boolean | void>;
		shouldReloadAccountManager?: () => boolean;
		reloadAccountManagerFromDisk?: () => Promise<void>;
		showToast?: (
			message: string,
			variant?: "info" | "success" | "warning" | "error",
		) => Promise<void>;
		onSelectionChanged?: (
			payload: RuntimeSelectionChangedPayload,
		) => Promise<void> | void;
	}) => Promise<boolean>;
	providerId: string;
	loadAccounts: () => Promise<TLoadedStorage>;
	saveAccounts: (storage: TSavedStorage) => Promise<void>;
	modelFamilies: readonly TModelFamily[];
	syncSelectedAccount?: (
		index: number,
		account: RuntimeSelectableAccount,
	) => Promise<boolean | void>;
	shouldReloadAccountManager?: () => boolean;
	reloadAccountManagerFromDisk?: () => Promise<void>;
	showToast?: (
		message: string,
		variant?: "info" | "success" | "warning" | "error",
	) => Promise<void>;
	onSelectionChanged?: (
		payload: RuntimeSelectionChangedPayload,
	) => Promise<void> | void;
	logDebug: (message: string) => void;
	pluginName: string;
}) {
	return async (input: { event: { type: string; properties?: unknown } }) => {
		try {
			const handled = await deps.handleAccountSelectEvent({
				event: input.event,
				providerId: deps.providerId,
				loadAccounts: deps.loadAccounts,
				saveAccounts: deps.saveAccounts,
				modelFamilies: deps.modelFamilies,
				syncSelectedAccount: deps.syncSelectedAccount,
				shouldReloadAccountManager: deps.shouldReloadAccountManager,
				reloadAccountManagerFromDisk: deps.reloadAccountManagerFromDisk,
				showToast: deps.showToast,
				onSelectionChanged: deps.onSelectionChanged,
			});
			if (handled) return;
		} catch (error) {
			deps.logDebug(
				`[${deps.pluginName}] Event handler error: ${error instanceof Error ? error.message : String(error)}`,
			);
		}
	};
}

export async function handleRuntimeEvent(params: {
	input: { event: { type: string; properties?: unknown } };
	providerId: string;
	modelFamilies: readonly ModelFamily[];
	loadAccounts: () => Promise<AccountStorageV3 | null>;
	saveAccounts: (storage: AccountStorageV3) => Promise<void>;
	syncSelectedAccount?: (
		index: number,
		account: RuntimeSelectableAccount,
	) => Promise<boolean | void>;
	shouldReloadAccountManager?: () => boolean;
	reloadAccountManagerFromDisk?: () => Promise<unknown>;
	showToast?: (
		message: string,
		variant?: "info" | "success" | "warning" | "error",
	) => Promise<void>;
	onSelectionChanged?: (
		payload: RuntimeSelectionChangedPayload,
	) => Promise<void> | void;
	logDebug: (message: string) => void;
	pluginName: string;
}): Promise<void> {
	try {
		await handleAccountSelectEvent({
			event: params.input.event,
			providerId: params.providerId,
			loadAccounts: params.loadAccounts,
			saveAccounts: params.saveAccounts,
			modelFamilies: params.modelFamilies,
			syncSelectedAccount: params.syncSelectedAccount,
			shouldReloadAccountManager: params.shouldReloadAccountManager,
			reloadAccountManagerFromDisk: params.reloadAccountManagerFromDisk,
			showToast: params.showToast,
			onSelectionChanged: params.onSelectionChanged,
		});
	} catch (error) {
		params.logDebug(
			`[${params.pluginName}] Event handler error: ${error instanceof Error ? error.message : String(error)}`,
		);
	}
}

File: crates/runtime/src/event_handler.rs
//! Port of `lib/runtime/event-handler.ts` — the never-throwing wrapper that
//! wires a runtime plugin event into the account-select handler
//! (spec 10 §19).

use crate::account_selection::{AccountSelectEventDeps, RuntimeEvent, handle_account_select_event};

/// Extra wiring for [`handle_runtime_event`] beyond the select-event deps.
pub struct RuntimeEventParams<'a> {
    pub plugin_name: &'a str,
    pub log_debug: &'a mut dyn FnMut(&str),
}

/// TS `handleRuntimeEvent(params)` — delegates to
/// [`handle_account_select_event`]; errors are debug-logged as
/// `"[{plugin}] Event handler error: {msg}"` and never propagated.
pub async fn handle_runtime_event<D: AccountSelectEventDeps>(
    event: &RuntimeEvent,
    deps: &mut D,
    params: &mut RuntimeEventParams<'_>,
) {
    if let Err(error) = handle_account_select_event(event, deps).await {
        (params.log_debug)(&format!(
            "[{}] Event handler error: {}",
            params.plugin_name,
            error.message()
        ));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use cma_core::errors::CodexError;
    use cma_core::schemas::account_storage::AccountStorageV3;
    use serde_json::json;

    /// Deps whose save always fails — proves the wrapper swallows errors.
    struct FailingDeps;

    impl AccountSelectEventDeps for FailingDeps {
        fn provider_id(&self) -> &str {
            "openai"
        }
        async fn load_accounts(&mut self) -> Result<Option<AccountStorageV3>, CodexError> {
            let mut storage = AccountStorageV3::empty();
            storage.accounts.push(
                cma_core::schemas::account_storage::AccountMetadataV3::new("rt", 1, 1),
            );
            Ok(Some(storage))
        }
        async fn save_accounts(&mut self, _: &AccountStorageV3) -> Result<(), CodexError> {
            Err(CodexError::new("disk full"))
        }
        fn has_cached_account_manager(&self) -> bool {
            false
        }
        async fn sync_codex_cli_active_selection_for_index(
            &mut self,
            _: usize,
        ) -> Result<(), CodexError> {
            Ok(())
        }
        fn set_last_codex_cli_active_sync_index(&mut self, _: usize) {}
        async fn reload_account_manager_from_disk(&mut self) -> Result<(), CodexError> {
            Ok(())
        }
        async fn show_toast(&mut self, _: &str, _: &str) -> Result<(), CodexError> {
            Ok(())
        }
    }

    #[tokio::test]
    async fn swallows_errors_and_logs_debug_line() {
        let mut logged = Vec::new();
        let mut log = |m: &str| logged.push(m.to_string());
        let event = RuntimeEvent {
            event_type: "account.select".to_string(),
            properties: Some(json!({"index": 0})),
        };
        handle_runtime_event(
            &event,
            &mut FailingDeps,
            &mut RuntimeEventParams {
                plugin_name: "codex-multi-auth",
                log_debug: &mut log,
            },
        )
        .await;
        assert_eq!(logged.len(), 1);
        assert!(
            logged[0].starts_with("[codex-multi-auth] Event handler error: "),
            "{}",
            logged[0]
        );
        assert!(logged[0].contains("disk full"));
    }
}
